packages/patterns/src/components/create-pattern-modal.js

```diff
@@ -74,14 +74,39 @@ export function CreatePatternModalContents( {
 
 	const { categoryMap, findOrCreateTerm } = useAddPatternCategory();
 
+	const form = useForm( {
+		title: {
+			validation: {
+				showErrorOnlyWhenDirty: true,
+				callback: ( data ) => {
```

Comment:
I feel this function should ideally be "absorbed" in a declarative way ultimately by having things like minLengh, maxLength, isRequired... in the field definition maybe.

Reply:
This means that what I wrote in the PR description isn't completely true:
```
The primary consideration is that fields should contain only UI logic, while business logic should reside in the consumer codebase. Consequently, the isValid callback should not be implemented at the field level.
```

I would avoid implementing validation on the field level, mostly I feel that different consumers could need different validation.

Reply:
Regarding the “declarative way,” we should support a generic function, allowing consumers to choose their preferred validation library.

However, we could use a declarative approach once Gutenberg adopts its own validation library (see: https://github.com/WordPress/gutenberg/discussions/67324).

Additionally, it’s important to note that implementing validation at the field level restricts validation to that specific field. It’s common to encounter scenarios where a field’s valid status depends on the value of another field.

Reply:
There are two questions:

 - Should it be field level, for and given the discussion in #67324 and the relationship with JSON schema, I think the field level would have my preference for now. We probably need more advanced multi-field validation later but that's for another day :P 
 - Should we start by the declarative approach or the escape hatch approach. I think for me this is still unclear to answer right now. If we go declarative, we need to try to match JSON schema. The con is it involves a bit more "architecture" work on the framework side.

Reply:
Doesn't set validation at the field level prevents consumers from applying custom validation to generic fields like “text” or “date”? They will need to create custom fields instead.

Reply:
> Doesn't set validation at the field level prevents consumers from applying custom validation to generic fields like “text” or “date”? They will need to create custom fields instead.

I don't understand how it would prevent anything? Here's a rough idea how I see things

```js
const fields = [ {
   id: 'title',
   type: 'string',
   minLength: 10
   maxLength: 50
   isRequired: true
} ]
```

Reply:
> Errors that can't be linked to a field probably come from server or before reaching the form, and those might be related to the form UI I think.

I don't know yet how it would look, I need use-cases to be able to think about it first but yeah I think forms can probably have global errors that are not directly attached to fields

Reply:
> I don't understand how it would prevent anything? Here's a rough idea how I see things

Oooh, ok. I was convicted that you were referring to putting validation constraint in [the field declaration](https://github.com/WordPress/gutenberg/blob/7b8acfa1fd1474f52fb30d499866c96e977826e8/packages/dataviews/src/field-types/text.tsx#L22-L28). Sorry for the confusion.

We could follow this approach, especially given the discussion in [WordPress Gutenberg #67324](https://github.com/WordPress/gutenberg/discussions/67324). However, we should also allow users to define custom callbacks if they prefer a different validation library or need to perform complex validations, such as multi-field validation.

The key point is to define the validation state on the consumer side and pass it to the dataform library. This ensures that both the library and the consumer can update the state as needed.

Reply:
> Oooh, ok. I was convicted that you were referring to putting validation constraint in [the field declaration](https://github.com/WordPress/gutenberg/blob/7b8acfa1fd1474f52fb30d499866c96e977826e8/packages/dataviews/src/field-types/text.tsx#L22-L28). Sorry for the confusion.

Yes, that example is a field.

> We could follow this approach, especially given the discussion in https://github.com/WordPress/gutenberg/discussions/67324. However, we should also allow users to define custom callbacks if they prefer a different validation library or need to perform complex validations, such as multi-field validation.

Maybe, but my main point is that this is an edge case and we should treat it once we face it as otherwise, I'm not sure we should be providing the right solution.

Reply:
> Yes, that example is a field.

If we decide to move validation to the field declaration level, how can consumers add their vaildation constraints to generic fields such as text or date?